### DISCUSSION TITLE
Nacos service watcher 是否有潜在的服务泄漏？

### DISCUSSION THREAD

**Discussion by spacewander** [higress-group/higress#769] [Q&A]
https://github.com/alibaba/higress/blob/c9c7df78a9ad183f5bc38d2f7b34c68514d1a095/registry/nacos/v2/watcher.go#L386

unsubscribe 的时候，看上去 Cache 里面对应 service 生成的 ServiceEntry 并没有被删除，这样就只能等待整个 watcher 更新时才会被删掉？

另外，
https://github.com/alibaba/higress/blob/c9c7df78a9ad183f5bc38d2f7b34c68514d1a095/registry/nacos/v2/watcher.go#L415
当 updateCacheWhenEmpty 为真时，实际上 ServiceEntry 并不会被删除，需要等待整个 watcher 更新时才会被真正删掉。

**Comment by johnlanni** [comment]
@spacewander 感谢关注，discussion没及时看不好意思，后面可以直接ping我。

这里是因为nacos在启动阶段时，higress去fetchAllServices可能会只拿到部分服务，所以不能直接在unsubscribe的时候就删除对应的ServiceEntry，因为实际上服务实例可能还在。

一般场景下服务数会趋向一个常量，这里的影响还好。

**Reply by spacewander** [reply]
明白了。所以这里是因为nacos有时不会返回全量服务，所以做了软删除？

**Reply by johnlanni** [reply]
是的
